### DISCUSSION TITLE
MetaEdit: Quick edit & add for YAML properties and Dataview inline fields

### DISCUSSION THREAD

**Discussion by chhoumann** [SilentVoid13/Templater#174] [Templates Showcase]
You are free to still use the template. However, I will no longer provide support for it.
Instead, please use the plugin: [https://github.com/chhoumann/MetaEdit](https://github.com/chhoumann/MetaEdit)

GET IT HERE: https://gist.github.com/chhoumann/2a90b203a4faaa35971a83f9e7f29d99

![O2OoLIPLes](https://user-images.githubusercontent.com/29108628/116616397-f7798400-a93c-11eb-8e3d-634e45bf3b68.gif)

If there are any bugs, let me know. 😄

This is pretty neat to use when combined with [hotkeys for templates](https://github.com/Vinzent03/obsidian-hotkeys-for-templates).

**Comment by chhoumann** [comment]
BUGFIX: In the previous version I forgot to check if the element was in YAML format or not. This has been fixed.

**Comment by chhoumann** [comment]
Update: Now you can add a new Dataview inline field whereever you please. You can also add new Yaml properties.

Dataview fields will be added above the selected line.
Yaml properties will simply be appended to the top of your property list.

**Comment by GitMurf** [comment]
Working great both updating and adding new YAML properties. The only other feature I would love to see, which I know is more complicated, would be adding a new value (a person may have multiple employees that are `Direct Reports`) instead of updating/replacing. Similar to this screenshot if I have two values in an array for the YAML. It would need to:

1) detect whether already have multiple values. Currently your script would put in the input box as placeholder "this is value,second one" because there was already multiple values. I suppose you could split() on comma separated so the user could just add an item after the default input in this situation.

2) if you decide to simply "Add a value" instead of "replace/update" then script detects whether a value for the property is already there and then turns the result that you modify the file with into an array of values like: `Property: ['first value', 'second value']` as opposed to currently which you would have: `Property: first value`.

Thoughts?

![image](https://user-images.githubusercontent.com/64155612/116630939-16f6b980-a909-11eb-804e-361e8049e87b.png)

**Comment by chhoumann** [comment]
Here's another update. I've tried to make some beneficial features that would help you automate your workflow.
That's why I've added these three things through customizable user settings:
- Ignored Properties
- Auto Props
- Multi-Value Mode (see @GitMurf comment above)

Starting in reverse order...
## Multi-Value Mode
This changes how you edit values. You will be prompted to either update the existing value or add a new value, which would change the property to an array/vector.
There are three options: 0 = off. 1 = on. 2 = on for certain properties.
![image](https://user-images.githubusercontent.com/29108628/116765023-84970880-aa23-11eb-845a-aa27a147ea0b.png)

## Auto Props
Have a defined set of values for specific properties? Here you can set them. You'll be prompted to set the value to one of those when you try to edit a property with that name. This is, of course, also customizable.
![LPBglgUtwt](https://user-images.githubusercontent.com/29108628/116765158-2c143b00-aa24-11eb-9e20-1bdcacaf842e.gif)

## Ignored Properties
These properties will not be shown in the suggester. This is for those properties that you know won't change.
![2ScpjYwGEm](https://user-images.githubusercontent.com/29108628/116765277-a6dd5600-aa24-11eb-958b-295d86d36e8a.gif)

**Comment by chhoumann** [comment]
2021-05-01 Part 2 - Smoother UX when adding and editing properties. 

When creating a new property with an autoprop name, you'll be prompted to select the autoprop value.
When editing in multivalue, options have been reordered for convenience.

**Comment by chhoumann** [comment]
# 2021-05-01 Part 3
NEW: 
**Progress Props**: (semi-)Automatically update task count, completed count, and incomplete count.
Demo:
![J9lIuwRg6L](https://user-images.githubusercontent.com/29108628/116795596-6c33f600-aad6-11eb-8482-c6cf7453edd7.gif)


Fixed bug where props and fields with intersecting names would be replaced by one of them.
Fixed bug for YAML properties in MultiValue Mode where they 1. wouldn't vectorize and 2. wouldn't get updated.

**Comment by chhoumann** [comment]
And just sneaked in some error handling for Progress Properties :)

**Comment by chhoumann** [comment]
Bugfix: if there are YAML delimiters but no properties it still adds delimiters

**Comment by ab-ll** [comment]
thats great, could you also get it to create new data fields depending on the action / include from other files i.e. if I have a status - then select complete; then it creates a completion date time and populates it and maybe adds a completion comments section to the document?

**Reply by chhoumann** [reply]
I mean, it's not that it isn't possible, but I think it would be really hard to make something that every user could benefit from - or, at least, not be impacted by.

If I can find a solution that works, I'll add it. But for now, I'd recommend just editing the script to fit your needs. In this case, I'd probably make a check for when it adds a certain property, then it could also perform the other actions.

**Reply by ab-ll** [reply]
👍

**Comment by chhoumann** [comment]
2021-05-04 Fixed bug that stopped the user from editing properties. Fixed bug with empty properties in MultiValue Mode.

**Comment by chhoumann** [comment]
2021-05-05 Fixed bug where you would be prompted to edit a variable even when you selected another option.

**Comment by Rhydderch** [comment]
Hello,
This looks amazing! I especially love the fact you have entered default values for specific fields!
However, I am a bit confused on how to use this script.
1. Where do I need to save the file and in which format (a note in templates folder or a js script under scripts folder) ?
2. How do I use it/call the function (explain it like I am 5) ? :)

**Reply by chhoumann** [reply]
Hey!
Thank you. :)

@GitMurf made a great video showcasing how you install scripts like this one:
https://github.com/SilentVoid13/Templater/discussions/187

You use it by 'inserting' the template.

**Reply by Rhydderch** [reply]
I got it working! Thank you :))
Your script + keyboard shortcuts = Yaml heaven

The script works but doesn't work fully on my end:
- I can add yaml fields and dataview fields (the prompts show up nicely)
- But I cannot edit yaml or dataview fields. When I click on "enter" to edit a field, the prompt doesn't appear and instead a blank space is inserted.
Do you have an idea what the problem might be?

**Reply by chhoumann** [reply]
Good to hear!

The bug is pretty odd. Never seen it before...
But you could try installing the MetaEdit plugin instead. I've fixed lots of weird bugs since transforming the script to an actual plugin.
The latest version is here: [https://github.com/chhoumann/MetaEdit/releases/tag/1.1.0](https://github.com/chhoumann/MetaEdit/releases/tag/1.1.0)

And installation instructions (because it's not in the community plugin broswer yet):
[https://github.com/chhoumann/MetaEdit](https://github.com/chhoumann/MetaEdit)

**Comment by MaksimZinovev** [comment]
@chhoumann Great plugin! Exactly what I was looking for.
How can I buy you a coffee?  😀

**Reply by chhoumann** [reply]
Hey! Thank you for asking! Right here https://www.buymeacoffee.com/chhoumann 😄

**Comment by SilentVoid13** [comment]
As specified by the author @chhoumann, you should now use the dedicated plugin https://github.com/chhoumann/MetaEdit for this.
Issues / Questions about this should now be asked on the MetaEdit github.
I'll now lock this discussion, glad that Templater helped launching a new awesome plugin, great work @chhoumann!
